hasktorch/src/Torch/Tensor.hs

```diff
@@ -109,16 +134,123 @@ mkScalarTensor v opts = unsafePerformIO $ do
 
 instance TensorLike Double where
   -- XXX: This implicit cast to float is very meh, but I don't have any better ideas
   --      This is our default dtype, so it's the most convenient thing we can do
   asTensor v = mkScalarTensor @Float (realToFrac v) float_opts
+  asValue t = realToFrac $ mkScalarValue @Float t
+
+--------------------------------------------------------------------------------
+-- List <-> Tensor promotion
+--------------------------------------------------------------------------------
+
+mkTensor :: (Storable a) => [a] -> [Int] -> TensorOptions -> Tensor
+mkTensor vs dims opts = unsafePerformIO $ do
+  t <- ((cast2 LibTorch.empty_lo) :: [Int] -> TensorOptions -> IO Tensor) dims opts
+  ptr <- ((cast1 ATen.tensor_data_ptr) :: Tensor -> IO (Ptr ())) t
+  forM_ (zip [0..] vs) $ \(i,v) ->
+    pokeElemOff (castPtr ptr) i v
+  return t
+
+mkValue :: (Storable a) => Tensor -> [a]
+mkValue t = unsafePerformIO $ do
+  let dims = shape t
+  ptr <- ((cast1 ATen.tensor_data_ptr) :: Tensor -> IO (Ptr ())) t
+  forM [0..((product dims)-1)] $ \i ->
+    peekElemOff (castPtr ptr) i
+
+split1d :: [Int] -> [a] -> [[a]]
+split1d _ [] = []
+split1d ns xs =
+  let n = head $ reverse ns
+      (x,y) = splitAt n xs
+  in x:(split1d ns y)
+
+split2d :: [Int] -> [a] -> [[[a]]]
+split2d _ [] = []
+split2d ns xs =
+  let (n0:n1:_) = reverse ns
+  in flip map (split1d [n0] xs) $ split1d [n1]
+
+instance TensorLike [Integer] where
+  asTensor v = mkTensor @Int64 (map fromIntegral v) [length v] int64_opts
+  asValue t = map fromIntegral $ mkValue @Int64 t
+
+instance TensorLike [Double] where
+  asTensor v = mkTensor @Float (map realToFrac v) [length v] float_opts
+  asValue t = map realToFrac $ mkValue @Float t
+
+instance TensorLike [[Integer]] where
+  asTensor v = mkTensor @Int64 (map fromIntegral (concat v)) [length v, length (head v)] int64_opts
+  asValue t = split1d (shape t) $ map fromIntegral $ mkValue @Int64 t
+
+instance TensorLike [[Double]] where
+  asTensor v = mkTensor @Float (map realToFrac (concat v)) [length v, length (head v)] float_opts
+  asValue t = split1d (shape t) $ map realToFrac $ mkValue @Float t
+
+instance TensorLike [[[Integer]]] where
+  asTensor v = mkTensor @Int64 (map fromIntegral (concat (concat v))) [length v, length (head v), length (head (head v))] int64_opts
+  asValue t = split2d (shape t) $ map fromIntegral $ mkValue @Int64 t
+
+instance TensorLike [[[Double]]] where
+  asTensor v = mkTensor @Float (map realToFrac (concat (concat v))) [length v, length (head v), length (head (head v))] float_opts
+  asValue t = split2d (shape t) $ map realToFrac $ mkValue @Float t
+
+---------------------------------------------------------------------------------------
+-- Tensor without variable. This is the same as torch.LongTensor or torch.FloatTensor.
+---------------------------------------------------------------------------------------
+
+class ConstTensorLike a where
```

Comment:
I'm not sure why do we even have this? The fact that a tensor can or cannot be a variable is really an implementation detail and it will be removed in the near future (every tensor will be a variable).

Reply:
When we use a tensor with variable, `sparse_coo_tensor` throws following exception.

```
Exception: Must not create a new variable from a variable, use its .data() (make_variable_consuming at libtorch/include/torch/csrc/autograd/variable.h:594)
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7ff8c6279441 inlibtorch/lib/libc10.so)
frame #1: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x2a (0x7ff8c6278d7a in libtorch/lib/libc10.so)
frame #2: .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/spec/spec() [0xaa9105]
frame #3: .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/spec/spec() [0xaa628a]
frame #4: .stack-work/dist/x86_64-linux/Cabal-2.4.0.1/build/spec/spec() [0xa86959]
; type: c10::Error
```

Reply:
I want to remove ConstTensor, if possible.